openhands/runtime/impl/action_execution/action_execution_client.py

```diff
@@ -0,0 +1,289 @@
+import os
+import tempfile
+import threading
+from abc import abstractmethod
+from pathlib import Path
+from typing import Any
+from zipfile import ZipFile
+
+import requests
+
+from openhands.core.config import AppConfig
+from openhands.core.exceptions import (
+    AgentRuntimeTimeoutError,
+)
+from openhands.events import EventStream
+from openhands.events.action import (
+    ActionConfirmationStatus,
+    BrowseInteractiveAction,
+    BrowseURLAction,
+    CmdRunAction,
+    FileEditAction,
+    FileReadAction,
+    FileWriteAction,
+    IPythonRunCellAction,
+)
+from openhands.events.action.action import Action
+from openhands.events.observation import (
+    ErrorObservation,
+    NullObservation,
+    Observation,
+    UserRejectObservation,
+)
+from openhands.events.serialization import event_to_dict, observation_from_dict
+from openhands.events.serialization.action import ACTION_TYPE_TO_CLASS
+from openhands.runtime.base import Runtime
+from openhands.runtime.plugins import PluginRequirement
+from openhands.runtime.utils.request import send_request
+
+
+class ActionExecutionClient(Runtime):
+    """Base class for runtimes that interact with the action execution server.
+
+    This class contains shared logic between DockerRuntime and RemoteRuntime
+    for interacting with the HTTP server defined in action_execution_server.py.
+    """
+
+    def __init__(
+        self,
+        config: AppConfig,
+        event_stream: EventStream,
+        sid: str = 'default',
+        plugins: list[PluginRequirement] | None = None,
+        env_vars: dict[str, str] | None = None,
+        status_callback: Any | None = None,
+        attach_to_existing: bool = False,
+        headless_mode: bool = True,
+    ):
+        self.session = requests.Session()
+        self.action_semaphore = threading.Semaphore(1)  # Ensure one action at a time
+        self._runtime_initialized: bool = False
+        self._vscode_token: str | None = None  # initial dummy value
+        super().__init__(
+            config,
+            event_stream,
+            sid,
+            plugins,
+            env_vars,
+            status_callback,
+            attach_to_existing,
+            headless_mode,
+        )
```

Comment:
Most thing LGTM! Do we keep the `LogStreamer` log in the client or still in each runtime implementation?

Reply:
Oh does the LogStreamer use the API now, instead of `docker logs`? If so we can centralize it!

Reply:
Seems like no--we need different log logic for docker, remote, modal, etc.

Mostly it seems like LogStreamer is a convenience when running locally, so only DockerRuntime has it for now. I removed the Modal and Runloop implementations since they're unused and hard to maintain